security/pkg/k8s/controller/workloadsecret.go

```diff
@@ -488,45 +493,19 @@ func (sc *SecretController) scrtUpdated(oldObj, newObj interface{}) {
 
 	_, waitErr := sc.certUtil.GetWaitTime(scrt.Data[CertChainID], time.Now(), sc.minGracePeriod)
 
-	caCert, _, _, rootCertificate := sc.ca.GetCAKeyCertBundle().GetAllPem()
-
-	// Check if root certificate in key cert bundle is not up-to-date. With mutiple
-	// Citadel deployed in Istio, the root certificate in istio-ca-secret could be
-	// rotated by any Citadel and become newer than the one in local key cert bundle.
-	if !bytes.Equal(rootCertificate, scrt.Data[RootCertID]) {
-		caSecret, scrtErr := sc.caSecretController.LoadCASecretWithRetry(CASecret,
-			sc.istioCaStorageNamespace, 100*time.Millisecond, 5*time.Second)
-		if scrtErr != nil {
-			k8sControllerLog.Errorf("Fail to load CA secret %s:%s (error: %s), skip updating secret %s",
-				sc.istioCaStorageNamespace, CASecret, scrtErr.Error(), name)
-			return
-		}
-		// The CA cert from istio-ca-secret is the source of truth. If CA cert
-		// in local keycertbundle does not match the CA cert in istio-ca-secret,
-		// reload root cert into keycertbundle.
-		if !bytes.Equal(caCert, caSecret.Data[caCertID]) {
-			k8sControllerLog.Warn("CA cert in KeyCertBundle does not match CA cert in " +
-				"istio-ca-secret. Start to reload root cert into KeyCertBundle")
-			var err error
-			rootCertificate, err = util.AppendRootCerts(caSecret.Data[caCertID], sc.rootCertFile)
-			if err != nil {
-				k8sControllerLog.Errorf("failed to append root certificates: %s", err.Error())
-				return
-			}
-			if err := sc.ca.GetCAKeyCertBundle().VerifyAndSetAll(caSecret.Data[caCertID],
-				caSecret.Data[caPrivateKeyID], nil, rootCertificate); err != nil {
-				k8sControllerLog.Errorf("failed to reload root cert into KeyCertBundle (%v)", err)
-				return
-			}
-			k8sControllerLog.Info("Successfully reloaded root cert into KeyCertBundle.")
-		}
+	caCertInMem, _, _, rootCertInMem := sc.ca.GetCAKeyCertBundle().GetAllPem()
+	syncedRootCert, err := sc.tryToSyncKeyCertBundle(rootCertInMem, caCertInMem)
```

Comment:
Please add this condition check:
if !bytes.Equal(rootCertificate, scrt.Data[RootCertID]) {
    syncedRootCert, err := sc.tryToSyncKeyCertBundle(rootCertInMem, caCertInMem)
}
(You need to update the names as well, no need to say "ImMem" or "synced" since the following code doesn't care about it anymore)

Otherwise you may always have a window of discrepancy between InMem KCB and the secret, which causes you to wrongly update the WL secret.

Reply:
Done. Thanks.